teps/0004-tekton-integrations-annotations-statuses.md

```diff
@@ -0,0 +1,689 @@
+installation to allow users to customize their installation.
+
+As a recommendation, integrations should rely on some form of reconciler in
+order to reliably process events, but it's also fine to rely on a push
+mechanism, or even rely on PipelineTasks to handle and annotate integration
+status data.
+
+#### Annotation-less integrations
+
+Defining annotations is an option, not a requirement. If you want an integration
+to apply to all Runs with no configuration needed (or provided by another source
+like a ConfigMap), that is okay. Integrations of this type will still able to
+use status fields.
+
+This is useful for use cases like webhook/pubsub notifications for all Run
+events.
+
+#### Triggers
+
+While this proposal mainly focuses on integrations with Pipeline and Task Run
+resources, Triggers are another important aspect to handle incoming integration
+events.
+
+To start, we can treat triggers the same as normal user requests - users will
+need to set relevant annotations in the created Task/PipelineRun. This is
+similar to the existing user experience today of needing to configure
+integration Tasks in their pipeline.
+
+There are Trigger related features being discussed that would lead to
+incremental improvements:
+
+- ObjectMeta Templating (e.g. be able to template annotations on created objects
+  per trigger) ([#618](https://github.com/tektoncd/triggers/issues/618))
+- Integration Specific Interceptors / Bindings / Templates
+  ([#504](https://github.com/tektoncd/triggers/issues/504),
+  [#584](https://github.com/tektoncd/triggers/issues/584))
+
+While these are out of scope for this specific proposal, this proposal remains
+compatible with future changes to make configuring Triggers easier, since it is
+agnostic to the source of configuration.
+
+#### Notifications
+
+This relates heavily to existing work around
+[notifications](https://github.com/tektoncd/pipeline/issues/1740), and aims to
+build on it to support metadata for arbitrary integrations alongside the
+existing
+[Cloud Event notifications](https://github.com/tektoncd/pipeline/issues/2082).
+While this proposal is largely compatible with existing work, one change would
+be recommended - notably migrating `CloudEventDelivery` as an integration
+status.
+
+Cloud Events are effectively a common type of a specific integration, and would
+fit well into this model in order to support arbitrary integration data. The
+Pipeline controller would be able to continue to insert data just like any other
+integration, and would be able to continue to write existing CloudEventDelivery
+status data alongside IntegrationStatus data.
+
+Example:
+
+[CloudEventDelivery](https://pkg.go.dev/github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1?tab=doc#CloudEventDelivery):
+
+```yaml
+target: example.com
+status:
+  condition: Failed
+  sentAt: 123456789
+  error: "access denied"
+  retryCount: 1
+```
+
+IntegrationStatus:
```

Comment:
This is the first tmie we introduce this type / field — the detail about it is lower, so it feels a bit weird (because at that point of reading, I don't know what is this type, where it is and what's is "goal".

Reply:
Yeah, I agree this is weird. Rearranged this a bit to not reference the type exactly until later. Let me know if this makes more sense or if I should make more changes.